cpp/test/sg/fil_test.cu

```diff
@@ -122,6 +140,37 @@ __global__ void nan_kernel(float* data, const bool* mask, int len, float nan)
 
 float sigmoid(float x) { return 1.0f / (1.0f + expf(-x)); }
 
+void hard_clipped_bernoulli(
+  raft::random::Rng rng, float* d, std::size_t n_vals, float prob_of_zero, cudaStream_t stream)
+{
+  rng.uniform(d, n_vals, 0.0f, 1.0f, stream);
+  thrust::transform(
+    thrust::cuda::par.on(stream), d, d + n_vals, d, [=] __device__(float uniform_0_1) {
+      float truly_0_1 = fmin(fmax(uniform_0_1, 0.0f), 1.0f);
+      return truly_0_1 > prob_of_zero;
```

Comment:
What if `prob_of_zero == 0.0f` and `truly_0_1 == 0.0f`? In this case, it can still generate `0.0f`, though it shouldn't (as the probability of it is zero). You could try `truly_0_1 >= prob_of_zero` then, but it won't work for `prob_of_zero == 1.0f` (it would generate 1 if `truly_0_1 == 1.0f`, but it shouldn't generate any 1 at all).

An alternative approach would be to ensure that `truly_0_1` cannot be equal to e.g. 1, and adjust the condition accordingly. Or to make sure it can't be equal to 0.